jobserver/authorization/utils.py

```diff
@@ -1,11 +1,102 @@
 import itertools
 
+from .converters import org_converter, project_converter
 
-def has_permission(user, permission):
+
+def _validate_context(allowed_keys, context):
+    """
+    Ensure keys and types-of-values in context are correct
+
+    When passing in context to either has_permission or has_role we want to
+    help developers track down incorrect keys or incorrect values.
+
+    For example, given this call:
+
+        _validate_context(["project"], {"projects": project})
+
+    we want to tell the develoer that "projects" is an invalid key.
+
+    Similarly we want to make sure the contextual objects passed in are valid
+    for the keys used, take the following function call:
+
+        _validate_context(["project"], {"project": user})
+
+    We want to tell the developer that a "project" key must map to a Project
+    (model) instance.
+    """
+    if not context:
+        return
+
+    # checks based on the allowed list passed in
+    unknown_keys = set(context.keys()) - set(allowed_keys)
+    if unknown_keys:
+        msg = "Some invalid keys were used in the context:"
+
+        for key in unknown_keys:
+            msg += f"\n - {key}"
+
+        raise ValueError(msg)
+
+    # import here to avoid circular imports
+    from jobserver import models
+
+    type_map = {
+        "org": models.Org,
+        "project": models.Project,
+    }
+    incorrect_types = {}
+    for key, value in context.items():
+        model = type_map[key]
+        if not isinstance(value, model):
+            incorrect_types[key] = type(value)
+
+    if not incorrect_types:
+        return
+
+    msg = "Some invalid values were used in the context:"
+    for key, value in incorrect_types.items():
+        expected = type_map[key]
+        msg += (
+            f'\n - key "{key}" got a {value.__name__}, expected a {expected.__name__}'
+        )
+
+    raise ValueError(msg)
+
+
+def _get_roles(user, **context):
+    """
+    Build up a set of Roles from the User and context
+
+    A User's Roles can come from various locations, and be local or global,
```

Comment:
local to a project or organisation?

Reply:
There's section in the OFP on Local vs Global (can't deep link to it sorry 😬 https://github.com/opensafely-core/roadmap/discussions/13), does that clarify the difference between the two types?

To your specific question, local to both, and the expectation of the design is we'll have more objects which need "local" Roles as we move forwards.  It's the concept that a Role is tied to a very particular object, eg you might need the ProjectDeveloper role on ProjectA but we don't want you have that role on every Project.